### PR TITLE
Fix early guest buffer read and implement ALSA capture

### PR DESCRIPTION
**Pull request by epilys**
- (All backends) Fixed the problem with reading stale data from the guest buffers
- (All backends) Fixed guest app writing data as fast as possible / sound not stopping after PcmStop
- (ALSA) Fixed invalid period/buffer size for host in alsa backend
- (Crate) Added some documentation comments
- (ALSA) Finally, implemented ALSA capture (RX).

### CODE REVIEW

File: crates/sound/src/audio_backends/alsa.rs [hunk 1]
```diff
@@ -490,9 +497,9 @@ impl AlsaBackend {
                             st.params.format = request.format;
                             st.params.rate = request.rate;
                         }
+                        // Manually drop msg for faster response: the kernel has a timeout.
+                        drop(msg);
```
Comment (stefano-garzarella) [inline crates/sound/src/audio_backends/alsa.rs:616]: Is this change intended?
Maybe better to describe it in the commit or split it into another commit (we would have to rework the story, so not splitting it is fine with me).

File: crates/sound/src/device.rs [hunk 2]
```diff
@@ -388,42 +388,51 @@ impl VhostUserSoundThread {
             return Ok(true);
         }
 
+        // Instead of counting descriptor chain lengths, encode the "parsing" logic in
+        // an enumeration. Then, the compiler will complain about any unhandled
+        // match {} cases if any part of the code is changed. This makes invalid
+        // states unrepresentable in the source code.
         #[derive(Copy, Clone, PartialEq, Debug)]
-        enum TxState {
+        enum IoState {
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs:471]: This change is in preparation of Rx support, right?

File: crates/sound/src/device.rs [hunk 3]
```diff
@@ -388,42 +388,51 @@ impl VhostUserSoundThread {
...
             let message = Arc::new(IOMessage {
                 vring: vring.clone(),
                 status: VIRTIO_SND_S_OK.into(),
                 latency_bytes: 0.into(),
                 desc_chain: desc_chain.clone(),
-                descriptor: descriptors.last().cloned().unwrap(),
+                response_descriptor: descriptors
+                    .last()
+                    .cloned()
+                    .expect("Received IO request with an empty descriptor chain."),
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: Should we return an error to the guest instead of panicking?

File: crates/sound/src/device.rs [hunk 4]
```diff
@@ -372,10 +373,11 @@ impl VhostUserSoundThread {
         Ok(false)
     }
 
-    fn process_tx(
+    fn process_io(
         &self,
         vring: &VringRwLock,
         audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
+        write: bool,
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: What about using an enum?

File: crates/sound/src/device.rs [hunk 5]
```diff
@@ -472,37 +470,34 @@ impl VhostUserSoundThread {
...
+                                    .period_bytes,
                             ) as usize,
                             descriptor.len(),
                         )
                         .into());
                     }
-                    IoState::WaitingBufferForStreamId(_stream_id) => {
-                        buffers.push(Buffer::new(*descriptor, Arc::clone(&message)));
+                    IoState::WaitingBufferForStreamId(_) => {
+                        buffers.push(Buffer::new(*descriptor, Arc::clone(&message), !write));
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: One point more for the enum.
I'd suggest to uniform `write/read`, for example here we use `write` and `is_rx` in Buffer. It is a bit confusing IMHO.

File: crates/sound/src/device.rs [hunk 6]
```diff
@@ -566,7 +561,7 @@ impl VhostUserSoundBackend {
             threads,
             virtio_cfg: VirtioSoundConfig {
                 jacks: 0.into(),
-                streams: 1.into(),
+                streams: Le32::from(streams_no as u32),
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs:669]: Is this related?

File: crates/sound/src/audio_backends/alsa.rs (hunk 1 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs:616]: Yes, the commit's title is `sound: clean up and improve comments, definitions` (or did it end up somewhere else? I'll check)

Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs:616]: To specify, the clean up here is to improve readability, the block was for early dropping the streams mutex and it could need a double take if someone sees them separated while they are meant to be performed together, logically

File: crates/sound/src/device.rs (hunk 2 above)
Comment (epilys) [reply crates/sound/src/device.rs:471]: Yes but done as `sound: clean up and improve comments, definitions`

File: crates/sound/src/device.rs (hunk 3 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Oh yes, good point.

File: crates/sound/src/device.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Would certainly be more comprehensible, will do

File: crates/sound/src/device.rs (hunk 5 above)
Comment (epilys) [reply crates/sound/src/device.rs]: +1

File: crates/sound/src/device.rs (hunk 6 above)
Comment (epilys) [reply crates/sound/src/device.rs:669]: Because RX is implemented, the second stream that is already allocated above as input won't be visible to the guest if the `streams` count is 1

File: crates/sound/src/device.rs (hunk 3 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Converted into error

File: crates/sound/src/device.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Added a `Direction` enum

File: crates/sound/src/device.rs (hunk 5 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Changed with:

> Added a Direction enum

**Comment by epilys** [comment]
Addressed review comments and force pushed

**Comment by dorindabassey** [comment]
issuing aplay with alsa backend doesn't work sometimes, see logs when that happens
```
[2023-10-09T12:47:33Z TRACE vhost_user_sound::device] dropping ControlMessage PcmSetParams reply = VIRTIO_SND_S_OK
[2023-10-09T12:47:33Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmPrepare
[2023-10-09T12:47:33Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStart
[2023-10-09T12:47:43Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStop
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:47:43Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:47:43Z TRACE vhost_user_sound::device] dropping ControlMessage PcmRelease reply = VIRTIO_SND_S_OK
[2023-10-09T12:48:01Z TRACE vhost_user_sound::device] dropping ControlMessage PcmSetParams reply = VIRTIO_SND_S_OK
[2023-10-09T12:48:01Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmPrepare
[2023-10-09T12:48:01Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStart
[2023-10-09T12:48:11Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStop
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40030d0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003140 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a40031b0 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f80a4003060 }
[2023-10-09T12:48:11Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-09T12:48:11Z TRACE vhost_user_sound::device] dropping ControlMessage PcmRelease reply = VIRTIO_SND_S_OK
 
```

**Comment by dorindabassey** [comment]
Thesame is the case for capture, Issuing `arecord` with the alsa backend doesn't work sometimes. Tested this PR with the latest QEMU patches https://patchew.org/QEMU/20231009095937.195728-1-alex.bennee@linaro.org/
Screenshot from 2023-10-10 09-46-03


```
[2023-10-10T07:43:37Z TRACE vhost_user_sound::device] dropping ControlMessage PcmSetParams reply = VIRTIO_SND_S_OK
[2023-10-10T07:43:37Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmPrepare
[2023-10-10T07:43:37Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStart
[2023-10-10T07:43:48Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStop
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f4018002fb0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f4018002fb0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180030b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180030b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180030b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180030b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180031b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180031b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f40180031b0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f4018002eb0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f4018002eb0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f4018002eb0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::stream] dropping Output buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Output, message: 0x7f4018002eb0 }
[2023-10-10T07:43:48Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(0) }
[2023-10-10T07:43:48Z TRACE vhost_user_sound::device] dropping ControlMessage PcmRelease reply = VIRTIO_SND_S_OK
[2023-10-10T07:44:04Z TRACE vhost_user_sound::device] dropping ControlMessage PcmSetParams reply = VIRTIO_SND_S_OK
[2023-10-10T07:44:04Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmPrepare
[2023-10-10T07:44:04Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStart
[2023-10-10T07:44:14Z TRACE vhost_user_sound::device] returned OK for ctrl msg PcmStop
[2023-10-10T07:44:14Z TRACE vhost_user_sound::stream] dropping Input buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Input, message: 0x7f4018002fb0 }
[2023-10-10T07:44:14Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(1000) }
[2023-10-10T07:44:14Z TRACE vhost_user_sound::stream] dropping Input buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Input, message: 0x7f4018002eb0 }
[2023-10-10T07:44:14Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(1000) }
[2023-10-10T07:44:14Z TRACE vhost_user_sound::stream] dropping Input buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Input, message: 0x7f40180030b0 }
[2023-10-10T07:44:14Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(1000) }
[2023-10-10T07:44:14Z TRACE vhost_user_sound::stream] dropping Input buffer Buffer { bytes: 0, pos: 0, populated: false, direction: Input, message: 0x7f4018003120 }
[2023-10-10T07:44:14Z TRACE vhost_user_sound] dropping IOMessage VirtioSoundPcmStatus { status: Le32(32768), latency_bytes: Le32(1000) }
[2023-10-10T07:44:14Z TRACE vhost_user_sound::device] dropping ControlMessage PcmRelease reply = VIRTIO_SND_S_OK

```

**Comment by epilys** [comment]
@dorindabassey Cannot reproduce locally. Sound should start after PcmStart is handled, so: Can you try adding the following line, to wake up the alsa worker thread after receiving a PcmStart manually:

```diff
diff --git a/crates/sound/src/audio_backends/alsa.rs b/crates/sound/src/audio_backends/alsa.rs
index 7971832..4e1a00c 100644
--- a/crates/sound/src/audio_backends/alsa.rs
+++ b/crates/sound/src/audio_backends/alsa.rs
@@ -535,6 +535,7 @@ impl AlsaBackend {
                             );
                         }
                     }
+                    senders[stream_id].send(true).unwrap();
                 }
                 AlsaAction::Prepare(stream_id) => {
                     if stream_id >= streams_no {
```

Maybe the VM is slower or faster in speed and the I/O messages arrive at different times when it's not working.
